### PR TITLE
Feat: new Ad component

### PR DESCRIPTION
**Pull request by BernieSumption**
Replaces the previous native and web implementations of the Ad component with a single cross platform implementation.

Major missing feature: targeting options for the page and ad slots are currently mocked in fixture files pending completion of work to generate this data correctly. #575 is the first part of this.

The code sharing mechanism that enables the cross platform component is DOMContext, which is a component that takes a function and data object and executes the function with an element and the data. On native this is done by serialising the function and data object to strings and injecting them into a WebView.

### CODE REVIEW

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/45a1dd8e843c5df2edccfd1918b851a0c418c0b1/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/77a2d4cecb6bc757344ec3fe4ef2673d7a9e82fc/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/53ba64172b175e8b1f3e910d3fbe6c4d52401f1a/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/22313c0bad5564d234b48b2e0fbd1ac8e26e4a12/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/47b436a42249d6edd1d875be32c0f234194639de/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/88c87e67d7f8e72bc75063d421119dc6c18d147b/dextrose-test/index.html

File: packages/ad/ad-init.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        Object.entries(data.slotOptions || {}).forEach(entry =>
+          slot.setTargeting(entry[0], entry[1])
+        );
+
+        renderComplete();
+      });
+    },
+
+    mount() {
```
Comment (craigbilner) [inline packages/ad/ad-init.js]: is this required?

File: packages/ad/__tests__/ad.test.js [hunk 2]
```diff
@@ -1,202 +1,71 @@
...
-import ReactTestUtils from "react-dom/test-utils";
+import Enzyme, { shallow } from "enzyme";
+import React16Adapter from "enzyme-adapter-react-16";
 
 import Ad from "../ad";
 
-describe("Native Ad test", () => {
+Enzyme.configure({ adapter: new React16Adapter() });
+
+jest.mock("../dom-context-harness", () => "mockHarness");
```
Comment (craigbilner) [inline packages/ad/__tests__/ad.test.js]: in general there is *a lot* of mocking going on. The initial implementation with the manager etc was meant to be based around DI a lot more to make testing easier and understanding of the code easier.

File: packages/ad/__tests__/check-serialisable-function.test.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import {
+  getSerialisableFunctionErrors,
+  reportErrors
+} from "./check-serialisable-function";
+
+describe("checkSerialisableFunction", () => {
+  it("does not find errors in funciton that does not use global variables", () => {
```
Comment (craigbilner) [inline packages/ad/__tests__/check-serialisable-function.test.js]: typo: function

File: packages/ad/__tests__/dom-context-harness.test.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  });
+
+  const fireLoadEventFor = source => {
+    Array.from(document.getElementsByTagName("script"))
+      .find(el => el.src === source)
+      .dispatchEvent(new window.Event("load"));
+  };
+
+  const makeHarness = args =>
+    _makeHarness({
```
Comment (craigbilner) [inline packages/ad/__tests__/dom-context-harness.test.js]: :horse_racing:

File: packages/ad/__tests__/ad.test.js (hunk 2 above)
Comment (BernieSumption) [reply packages/ad/__tests__/ad.test.js]: This isn't really mocking, it's for cleaning up snapshots. I'd love to have a general purpose solution to the issue. I have issues with snapshots, so often they generate thousand line monstrosities. These 3 lines prevent the transpiled source of the functions (complete with code coverage statements that change on every build), showing up in snapshots:

```jest.mock("../dom-context-harness", () => "mockHarness");
jest.mock("../webview-event-callback-setup", () => "mockErrorHandler");
jest.mock("../ad-init", () => () => {});```

And this one prevents an inline SVG doing the same:

```jest.mock("../placeholder", () => "Placeholder");```

Some more general way of controlling what lands in snapshots would be good. I'd start with a general rule that snapshots should never descend into components defined in other times-components packages.

File: packages/ad/ad-init.js (hunk 1 above)
Comment (BernieSumption) [reply packages/ad/ad-init.js]: It is required for testability - by having the init function return an object and then calling a method on the object, it's possible to install spies on other members before calling the method.

However `mount()` is a bad name chosen when I thought that there was going to be an `unmount()` too, so I've renamed it to `execute()` for consistency with dom-context-harness.js

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/815cc0ede0c3d3008e2800ee26308fa242450feb/dextrose-test/index.html

File: packages/ad/dom-context.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          }
+        </style>
+        </head>
+        <body>
+          <div></div>
+          <script>
+            (${webviewEventCallbackSetup})(window);
+          </script>
+          <script>
+            var harness = (${makeHarness})({
```
Comment (craigbilner) [inline packages/ad/dom-context.js]: is the declaration needed?

File: packages/ad/webview-event-callback-setup.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        JSON.stringify({
+          isTngMessage: true,
+          type,
+          detail
+        }),
+        "*"
+      );
+    });
+  };
+  window.onerror = (msg, file, line, col, error) => {
```
Comment (craigbilner) [inline packages/ad/webview-event-callback-setup.js]: guessing there'll be no existing methods in the WebView we care about trashing?

File: packages/ad/__tests__/ad.test.js (hunk 2 above)
Comment (craigbilner) [reply packages/ad/__tests__/ad.test.js]: happy for you to say, "that's over-engineering" but would it be neater if ad took these components as "render props"?

**Comment by BernieSumption** [comment]
@craigbilner Implemented all these changes, except _happy for you to say, "that's over-engineering" but would it be neater if ad took these components as "render props"_. I implemented it and it looks worse than the `jest.mock` that it's supposed to improve on - Ad component being aware of internal implementation details of DOMContext, just so that it can override them in tests, and never using the feature in production. So I added a code comment explaining why the `jest.mock` statements are there.

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/0a6317212bcdb518821a7f98d68fb77e70e72033/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/81181359f033d741f1a1365d5dc1cc7933ab157c/dextrose-test/index.html

**Pull request merged**